pkg/volume/util/atomic_writer.go

```diff
@@ -0,0 +1,445 @@
+//      the real data:
+//        <target-dir>/..data          -> ..data.12345678/
+//  6.  The current timestamped directory is detected by reading the data directory
+//      symlink
+//  7.  A symlink to the new timestamped directory ..data_tmp is created that will
+//      become the new data directory
+//  8.  The new data directory symlink is renamed to the data directory; rename is atomic
+//  9.  The sentinel file modification and access times are updated (file is created if it does not
+//      already exist)
+// 10.  Old paths are removed from the user-visible portion of the target directory
+// 11.  The previous timestamped directory is removed, if it exists
+func (w *AtomicWriter) Write(payload map[string][]byte) error {
+	cleanPayload, err := validatePayload(payload)
+	if err != nil {
+		glog.Errorf("%s: invalid payload: %v", w.logContext, err)
+		return err
+	}
+
+	pathsToRemove, err := w.pathsToRemove(cleanPayload)
+	if err != nil {
+		glog.Errorf("%s: error determining user-visible files to remove: %v", w.logContext, err)
+		return err
+	}
+
+	if should, err := w.shouldWritePayload(cleanPayload); err != nil {
+		glog.Errorf("%s: error determining whether payload should be written to disk: %v", w.logContext, err)
+		return err
+	} else if !should && len(pathsToRemove) == 0 {
+		glog.V(5).Infof("%s: no update required for target directory %v", w.logContext, w.targetDir)
+		return nil
+	}
+
+	tsDir, err := w.newTimestampDir()
+	if err != nil {
+		glog.V(5).Infof("%s: error creating new ts data directory: %v", w.logContext, err)
+		return err
+	}
+
+	if err = w.writePayloadToDir(cleanPayload, tsDir); err != nil {
+		glog.Errorf("%s: error writing payload to ts data directory %s: %v", w.logContext, tsDir, err)
+		return err
+	}
+
+	if err = w.createUserVisibleFiles(cleanPayload); err != nil {
+		glog.Errorf("%s: error creating visible symlinks in %s: %v", w.logContext, w.targetDir, err)
+		return err
+	}
+
+	_, tsDirName := filepath.Split(tsDir)
+	dataDirPath := path.Join(w.targetDir, dataDirName)
+	oldTsDir, err := os.Readlink(dataDirPath)
+	if err != nil && !os.IsNotExist(err) {
+		glog.Errorf("%s: error reading link for data directory: %v", w.logContext, err)
+		return err
+	}
+
+	newDataDirPath := path.Join(w.targetDir, newDataDirName)
+	if err = os.Symlink(tsDirName, newDataDirPath); err != nil {
+		os.RemoveAll(tsDir)
+		glog.Errorf("%s: error creating symbolic link for atomic update: %v", w.logContext, err)
+		return err
+	}
+
+	if err = os.Rename(newDataDirPath, dataDirPath); err != nil {
+		os.Remove(newDataDirPath)
+		os.RemoveAll(tsDir)
+		glog.Errorf("%s: error renaming symbolic link for data directory %s: %v", w.logContext, newDataDirPath, err)
+		return err
+	}
+
+	if err = w.touchSentinelFile(); err != nil {
+		glog.Errorf("%s: error touching sentinel file: %v", err)
+		return err
+	}
+
+	if err = w.removeUserVisiblePaths(pathsToRemove); err != nil {
```

Comment:
no, why?  The sentinal is the signal that the old dataset is dead.  Until
we touch the sentinal, the app could rightfully use any file in the
dataset, and any error we cause is a really bad thing.

On Tue, Feb 9, 2016 at 8:03 PM, Seth Jennings <notifications@github.com>
wrote:

> In pkg/volume/util/atomic_writer.go
> <https://github.com/kubernetes/kubernetes/pull/20397#discussion_r52412448>
> :
>
> > +		return err
> > +	}
> > +
> > +	if err = os.Rename(newDataDirPath, dataDirPath); err != nil {
> > +		os.Remove(newDataDirPath)
> > +		os.RemoveAll(tsDir)
> > +		glog.Errorf("%s: error renaming symbolic link for data directory %s: %v", w.logContext, newDataDirPath, err)
> > +		return err
> > +	}
> > +
> > +	if err = w.touchSentinelFile(); err != nil {
> > +		glog.Errorf("%s: error touching sentinel file: %v", err)
> > +		return err
> > +	}
> > +
> > +	if err = w.removeUserVisiblePaths(pathsToRemove); err != nil {
>
> should we remove the old user visible paths before touching the sentinel?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20397/files#r52412448>.
>